File: String/lcstring.py
def lcstring(str1, str2):
    lena = len(str1)
    lenb = len(str2)
    maxval = 0
    maxi = 0
    maxj = 0
    matrix = [[0] * (lena + 1) for i in range(lenb + 1)]
    for i in range(1, lenb + 1):
        for j in range(1, lena + 1):
            if str1[j - 1] == str2[i - 1]:
                matrix[i][j] = matrix[i - 1][j - 1] + 1
                if matrix[i][j] > maxval:
                    maxval = matrix[i][j]
                    maxi = i
                    maxj = j
    result = ""
    while (maxi > 0 and maxj > 0):
        result = str1[maxj - 1] + result
        maxi -= 1
        maxj -= 1
    return result

if __name__ == "__main__":
    a = "12345678999999987654321"
    b = "87654321"
    result = lcstring(a, b)
    print("stringA:\t%s\n stringB:\t%s\n lsctring:\t%s\n length:\t%s" % (a, b, result, len(result)));

File: String/lcstring.c
#include <string.h>
#include <stdlib.h>
#include <stdio.h>

void pr_matrix(int **matrix, int alen, int blen)
{
		for (int i = 0; i <= blen; i++)
		{
				for (int j = 0; j <= alen; j++)
				{
						printf("%d\t", matrix[i][j]);
				}
				printf("\n");
		}
}


char *lcstring(char *str1, char *str2)
{
		int alen = strlen(str1);
		int blen = strlen(str2);
		int maxi, maxj, currval, maxval = 0, k = 0;

		int **matrix = malloc(sizeof(int *) * (blen + 1));
		for (int i = 0; i <= blen; i++)
		{
				*(matrix + i) = malloc(sizeof(int) * (alen + 1));
				memset(*(matrix + i), 0, sizeof(int) * (alen + 1));
		}
		for (int i = 1; i <= blen; i++)
		{
				for (int j = 1; j <= alen; j++)
				{
						if (str1[j - 1] == str2[i - 1])
						{
								currval = matrix[i - 1][j - 1] + 1;
								if (currval > maxval)
									{ maxval = currval; maxi = i; maxj = j; }
								matrix[i][j] = currval;
						}
				}
		}
		//pr_matrix(matrix, alen, blen);

		char *s = malloc(sizeof(char) * (maxval + 1));
		k = maxval - 1;
		for (; maxi > 0 && maxj > 0; maxi--, maxj--)
		{
				s[k--] = str2[maxi - 1];
		}
		s[maxval] = '\0';
		return(s);
}

int main(int argc, char *argv[])
{
		char a[] = "tutorialhorizon";
		char b[] = "dynamictutorialProgramming";
		char *c = lcstring(a, b);
		printf("stringA:\t%s\nstringB:\t%s\nlcstring:\t%s\nlength: \t%ld\n", a, b, c, strlen(c));
		return(0);
}

